telink: tl321x: fix potential ram crash.

haiwentelink opened the pull request:
- seperate ilm and dram.
- put more sections to dram.

fengtai-telink commented on soc/riscv/riscv-privilege/telink_tlx/linker.ld:
```diff
@@ -25,7 +25,7 @@ MEMORY
 	RAMILM       (rwx) : ORIGIN = DT_REG_ADDR(DT_NODELABEL(ram_ilm)),              LENGTH = DT_REG_SIZE(DT_NODELABEL(ram_ilm))
 	#elif CONFIG_SOC_RISCV_TELINK_TL321X
-
+	RAM_DLM       (rwx) : ORIGIN = DT_REG_ADDR(DT_NODELABEL(ram_dlm)),              LENGTH = DT_REG_SIZE(DT_NODELABEL(ram_dlm))
```
The code format needs to be slightly adjust, need to delete one space

haiwentelink replied:
done

andriy-bilynskyy commented on soc/riscv/riscv-privilege/telink_tlx/linker.ld:
```diff
@@ -25,7 +25,7 @@ MEMORY
 	RAMILM       (rwx) : ORIGIN = DT_REG_ADDR(DT_NODELABEL(ram_ilm)),              LENGTH = DT_REG_SIZE(DT_NODELABEL(ram_ilm))
 	#elif CONFIG_SOC_RISCV_TELINK_TL321X
-
+	RAM_DLM       (rwx) : ORIGIN = DT_REG_ADDR(DT_NODELABEL(ram_dlm)),              LENGTH = DT_REG_SIZE(DT_NODELABEL(ram_dlm))
```
(rw) - we can't execute from DLM

haiwentelink replied:
we want to move more ram to dlm , or ilm is not enough , dlm can run ramcode .

andriy-bilynskyy replied:
Do you have any datasheet where is mentioned that code can be executed from DLM?

haiwentelink replied:
@andriy-bilynskyy  see the two pictures of the ticket http://192.168.48.49:8080/browse/MATTER-677 , for the details , i need to ask shanghai team to send the pdf file .

andriy-bilynskyy replied:
In the past in datasets we have:
- instruction local memory (ILM)
- data local memory (DLM)

names clear describe their purposes.
If both parts are capable to execute code and the only difference is retention feature
it's reasonable to rename them
See as example _boards/telink/tl7218/tl7218_retention.dts_
```
&ram_ilm_retention {
	reg = <0x00000000 0x00018000>;
};

&ram_ilm_nonretention {
	reg = <0x00018000 0x00028000>;
};
```
Also renaming should be applied to linker script.
Otherwise we got confused code which is near impossible to maintain.

haiwentelink replied:
@andriy-bilynskyy agree , will change .i also think it's reasonable.

haiwentelink replied:
@andriy-bilynskyy already changed ,can start review again .

andriy-bilynskyy commented on soc/riscv/riscv-privilege/telink_tlx/telink_tlx_regions.ld:
```diff
@@ -11,8 +11,10 @@
+		#define TELINK_RAM_CODE_REGION         RAM_DLM
+		#define TELINK_RAM_CODE_RET_REGION     RAM_DLM
+		#define TELINK_RAM_DATA_REGION         RAM_DLM
+		#define TELINK_RAM_CODE_BLE_REGION     RAM_DLM
```
Seems that's not possible to put executed code into DLM

haiwentelink replied:
dlm is not enough ,because the data section and noinit_ilm  are bigger than 32k , i have no choice to move this to dlm.

haiwentelink replied:
@andriy-bilynskyy  already changed , and use homepod mini can test pass . so you can start review again .

haiwentelink committed:
telink: tl321x: fix potential ram crash.

- seperate ilm and ilm_non_retention.
- put more sections to ilm_non_retention.

Signed-off-by: Haiwen Xia <haiwen.xia@telink-semi.com>

s07641069 commented:
Duplicated PR into develop_new branch (Zephyr 3.7) https://github.com/telink-semi/zephyr/pull/453
